blog/2021-05-29-using-rust.md

```diff
@@ -0,0 +1,96 @@
+- The Rust book was a great resource. It had a lot of examples and explanations for Rust fundamentals, as well a great insight into multithreading and concurrency.
+- It's a little harder to refactor and make huge changes compared to other languages, but it's a worthwhile tradeoff for a critical service in our opinion, as it makes the code more robust. It's also easier to reason with ownership rules once you get the hang of it, compared to compiler magic that some other languages have.
+
+Cons :
+
+- We faced some issues with key libraries not being well maintained. For example, we use AWS, and thus were using the [`rusoto`](https://github.com/rusoto/rusoto) library. Because this library had a dependency with Tokio `0.1.15`, we couldn't migrate to Tokio `1.x` for a really long time. We were able to do it later when `rusoto` was updated, but we still expected such a critical library to stay up to date. Things are looking good however, with AWS announcing that they are [working on an official SDK for Rust](https://aws.amazon.com/blogs/developer/a-new-aws-sdk-for-rust-alpha-launch/).
+- It's always a risk upgrading a lot of dependencies in the current state of rust, as you can almost always expect that some inter-dependency compatibility breaks. This was especially magnified in a large project such as this one.
+
+We also have some general takeaways and gotchas we encountered in this project:
+
+- _Tensorflow_ : We found that that the prebuilt tensorflow `C` bindings were not built/available for a large variety of GPU instances we use in AWS. This proved to be a little tedious to fix, as we had to manually compile tensorflow for the systems we use in production, without which we experienced slow inference times and model loading.
+- It's pretty hard and largely undocumented how to run ML model inference **in parallel** on GPUs, and we still have not fully explored that option.
+
+[^1]: The image hash is calculated and used to check for duplicates.
+[^2]: [Tokio](https://tokio.rs/) is an asynchronous runtime for the Rust programming language. A lot of languages have a built in async runtime. Rust allows you to choose whichever runtime you require. Tokio is the most popular option in the Rust ecosystem. Check out [this resource](https://rust-lang.github.io/async-book/08_ecosystem/00_chapter.html) for more insight into async and the rust async ecosystem!
+[^3]: Amazon [`SQS`](https://aws.amazon.com/sqs/) is a fully managed queue which we were using to distribute messages to different Rust service instances.
+[^4]: We used [`warp`](https://github.com/seanmonstar/warp) because of it's excellent tokio interoperability and flexible [`Filter`](https://docs.rs/warp/0.1.0/warp/trait.Filter.html) system.
```

Comment:
```suggestion
[^4]: We used [`warp`](https://github.com/seanmonstar/warp) because of its excellent tokio interoperability and flexible [`Filter`](https://docs.rs/warp/0.1.0/warp/trait.Filter.html) system.
```